eip712/messages.py

```diff
@@ -144,3 +149,139 @@ def signable_message(self) -> SignableMessage:
+
+    https://docs.metamask.io/guide/signing-data.html#sign-typed-data-v4
+
+    This code is ported from the Javascript "eth-sig-util" package.
+    """
+    encoded_types = ["bytes32"]
+    encoded_values = [hash_type(primary_type, types)]
+
+    def _encode_field(name, typ, value):
+        if typ in types:
+            if value is None:
+                return [
+                    "bytes32",
+                    "0x0000000000000000000000000000000000000000000000000000000000000000",
+                ]
+            else:
+                return ["bytes32", keccak_256(encode_data(typ, value, types)).digest()]
+
+        if value is None:
+            raise Exception(f"Missing value for field {name} of type {type}")
+
+        # Accept string bytes
+        if "bytes" in typ and isinstance(value, str):
+            value = HexBytes(value)
+
+        # Accept string uint and int
+        if "int" in typ and isinstance(value, str):
+            value = int(value)
+
+        if typ == "bytes":
+            return ["bytes32", keccak_256(value).digest()]
+
+        if typ == "string":
+            # Convert string to bytes.
+            value = value.encode("utf-8")
+            return ["bytes32", keccak_256(value).digest()]
+
+        if typ.endswith("]"):
+            # Array type
+            if value:
+                parsed_type = typ[: typ.rindex("[")]
+                type_value_pairs = [_encode_field(name, parsed_type, v) for v in value]
+                data_types, data_hashes = zip(*type_value_pairs)
+            else:
+                # Empty array
+                data_types, data_hashes = [], []
+
+            h = keccak_256(encode_abi(data_types, data_hashes)).digest()
+            return ["bytes32", h]
+
+        return [typ, value]
+
+    for field in types[primary_type]:
+        typ, val = _encode_field(field["name"], field["type"], data[field["name"]])
+        encoded_types.append(typ)
+        encoded_values.append(val)
+
+    return encode_abi(encoded_types, encoded_values)
+
+
+def eip712_encode(typed_data: Dict[str, Any]) -> List[bytes]:
+    """
+    Given a dict of structured data and types, return a 3-element list of
+    the encoded, signable data.
+
+      0: The magic & version (0x1901)
+      1: The encoded types
+      2: The encoded data
+    """
+    try:
+        parts = [
+            bytes.fromhex("1901"),
+            hash_struct("EIP712Domain", typed_data["domain"], typed_data["types"]),
+        ]
+        if typed_data["primaryType"] != "EIP712Domain":
+            parts.append(
+                hash_struct(
+                    typed_data["primaryType"],
+                    typed_data["message"],
+                    typed_data["types"],
+                )
+            )
+        return parts
+    except (KeyError, AttributeError, TypeError, IndexError) as exc:
+        raise ValueError(f"Not valid {typed_data}") from exc
+
+
+def eip712_encode_hash(typed_data) -> str:
```

Comment:
If there is a way to use only the methods from eth-account, please help me. I for the life of me can't get it to work, it is so confusing.